e2e/metrics/metrics.go

```diff
@@ -0,0 +1,196 @@
+package metrics
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/e2e/framework"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/prometheus/common/model"
+
+	"github.com/stretchr/testify/require"
+)
+
+type MetricsTest struct {
+	framework.TC
+	jobIDs       []string
+	prometheusID string
+	fabioID      string
+	fabioAddress string
+}
+
+func init() {
+	framework.AddSuites(&framework.TestSuite{
+		Component:   "Metrics",
+		CanRunLocal: true,
+		Cases: []framework.TestCase{
+			new(MetricsTest),
+		},
+	})
+}
+
+// Stand up prometheus to collect metrics from all clients and allocs,
+// with fabio as a system job in front of it so that we don't need to
+// have prometheus use host networking
+func (tc *MetricsTest) BeforeAll(f *framework.F) {
+
+	e2eutil.WaitForLeader(f.T(), tc.Nomad())
+	e2eutil.WaitForNodesReady(f.T(), tc.Nomad(), 1)
+
+	require := require.New(f.T())
+	err := tc.setUpPrometheus(f)
+	require.Nil(err)
+}
+
+// Clean up the target jobs after each test case, but keep fabio/prometheus
+// for reuse between the two test cases (Windows vs Linux)
+func (tc *MetricsTest) AfterEach(f *framework.F) {
+	if os.Getenv("NOMAD_TEST_SKIPCLEANUP") == "1" {
+		return
+	}
+	for _, jobID := range tc.jobIDs {
+		tc.Nomad().Jobs().Deregister(jobID, true, nil)
+	}
+	tc.jobIDs = []string{}
+	tc.Nomad().System().GarbageCollect()
+}
+
+// Clean up fabio/prometheus
+func (tc *MetricsTest) AfterAll(f *framework.F) {
+	if os.Getenv("NOMAD_TEST_SKIPCLEANUP") == "1" {
+		return
+	}
+	tc.tearDownPrometheus(f)
+}
+
+// TestMetricsLinux runs a collection of jobs that exercise alloc metrics.
+// Then we query prometheus to verify we're collecting client and alloc metrics
+// and correctly presenting them to the prometheus scraper.
+func (tc *MetricsTest) TestMetricsLinux(f *framework.F) {
+	clientNodes, err := e2eutil.ListLinuxClientNodes(tc.Nomad())
+	if err != nil {
+		f.T().Fatalf("could not list client nodes: %v", err)
+	}
+	if len(clientNodes) == 0 {
```

Comment:
should we potentially refactor this into a test helper?

Reply:
I liked the idea of this and took a crack at it, but it turns out I need the list of nodes that we query as part of that skip later on in every case I have it, so it adds rather than removes code to do it.

Reply:
I've at least dropped the `if err != nil` by moving it into a `require.Nil` though.